swe_roe/ex2.c

```diff
@@ -8,418 +8,461 @@ static char help[] = "Partial 2D dam break problem.\n";
+        vertices->num_edges[ivertex]++;
+      } else {
+        PetscInt offset           = vertices->cell_offsets[ivertex];
+        PetscInt index            = offset + vertices->num_cells[ivertex];
+        vertices->cell_ids[index] = p[i] - cStart;
+        vertices->num_cells[ivertex]++;
+      }
+    }
 
-  for (PetscInt icell = 0; icell < num_cells; icell++) {
-    cells->id[icell]            = icell;
-    cells->num_vertices[icell]  = num_vertices;
-    cells->num_edges[icell]     = num_edges;
-    cells->num_neighbors[icell] = num_neighbors;
+    PetscCall(DMPlexRestoreTransitiveClosure(dm, v, PETSC_FALSE, &pSize, &p));
   }
 
-  for (PetscInt icell = 0; icell <= num_cells; icell++) {
-    cells->vertex_offset[icell]   = icell * num_vertices;
-    cells->edge_offset[icell]     = icell * num_edges;
-    cells->neighbor_offset[icell] = icell * num_neighbors;
-  }
+  VecRestoreArray(coordinates, &coords);
 
   PetscFunctionReturn(0);
 }
 
-/// Allocates memory and initialize a RDyEdge struct
-/// @param [in] num_edges Number of edges
-/// @param [out] edges A RDyEdge struct that is allocated
+/// Destroys an RDyVertices struct, freeing its resources.
+/// @param [inout] vertices An RDyVertices struct whose resources will be freed
 ///
 /// @return 0 on success, or a non-zero error code on failure
-PetscErrorCode AllocateEdges(PetscInt num_edges, RDyEdge *edges) {
+PetscErrorCode RDyVerticesDestroy(RDyVertices vertices) {
   PetscFunctionBegin;
 
-  PetscInt num_cells = 2;
+  PetscCall(PetscFree(vertices.ids));
+  PetscCall(PetscFree(vertices.global_ids));
+  PetscCall(PetscFree(vertices.is_local));
+  PetscCall(PetscFree(vertices.num_cells));
+  PetscCall(PetscFree(vertices.num_edges));
+  PetscCall(PetscFree(vertices.cell_offsets));
+  PetscCall(PetscFree(vertices.edge_offsets));
+  PetscCall(PetscFree(vertices.cell_ids));
+  PetscCall(PetscFree(vertices.edge_ids));
+  PetscCall(PetscFree(vertices.points));
 
-  PetscCall(RDyAllocate_IntegerArray_1D(&edges->id, num_edges));
-  PetscCall(RDyAllocate_IntegerArray_1D(&edges->global_id, num_edges));
-  PetscCall(RDyAllocate_IntegerArray_1D(&edges->num_cells, num_edges));
-  PetscCall(RDyAllocate_IntegerArray_1D(&edges->vertex_ids, num_edges * 2));
-
-  PetscCall(RDyAlloc(num_edges * sizeof(PetscBool), &edges->is_local));
-  PetscCall(RDyAlloc(num_edges * sizeof(PetscBool), &edges->is_internal));
+  PetscFunctionReturn(0);
+}
 
-  PetscCall(RDyAllocate_IntegerArray_1D(&edges->cell_offset, num_edges + 1));
-  PetscCall(RDyAllocate_IntegerArray_1D(&edges->cell_ids, num_edges * num_cells));
+/// A struct of arrays storing information about edges separating mesh cells.
+/// The ith element in each array stores a property for edge i.
+typedef struct {
+  /// local IDs of edges in local numbering
+  PetscInt *ids;
+  /// global IDs of edges in local numbering
+  PetscInt *global_ids;
+
+  /// PETSC_TRUE iff edge is shared by locally owned cells, OR
```

Comment:
@donghuix , here's the comment you were talking about. I didn't get rid of any comments--just marked them up with Doxygen tags and elaborated.